code/__HELPERS/areas.dm

```diff
@@ -0,0 +1,86 @@
+// Gets an atmos isolated contained space
+// Returns an associative list of turf|dirs pairs
+// The dirs are connected turfs in the same space
+// break_if_found is a typecache of turf types to return false if found
+/proc/detect_room(turf/origin, list/break_if_found)
+	if(origin.blocks_air)
+		return list(origin)
+
+	. = list()
+	var/list/checked_turfs = list()
+	var/list/found_turfs = list(origin)
+	while(found_turfs.len)
+		var/turf/sourceT = found_turfs[1]
+		if(break_if_found[sourceT.type])
+			return FALSE
+		found_turfs.Cut(1, 2)
+		var/dir_flags = checked_turfs[sourceT]
+		for(var/dir in GLOB.alldirs)
+			if(dir_flags & dir) // This means we've checked this dir before, probably from the other turf
+				continue
+			var/turf/checkT = get_step(sourceT, dir)
+			if(!checkT)
+				continue
+			checked_turfs[sourceT] |= dir
+			checked_turfs[checkT] |= turn(dir, 180)
+			.[sourceT] |= dir
+			.[checkT] |= turn(dir, 180)
+			var/static/list/diagonal_cache = GLOB.diagonals
+			if((dir in diagonal_cache) || checkT.blocks_air || !CANATMOSPASS(sourceT, checkT))
```

Comment:
*family feud angry noise*
try again.

Reply:
The other option is to do diagonal_cache as a associative list of dirs as text and compare to that. Not sure what the costs there is. I can't think of any other methods at the moment but it's 6 am

Reply:
```
var/static/list/diagonal_cache = list("[NORTH]" = TRUE, "[SOUTH]" = TRUE, "[EAST]" = TRUE, "[WEST]" = TRUE)
if((diagonal_cache["[dir]"]...)
```
Should work fine, as for performance, one: in sucks ass so I doubt some string operations would offset the benefit of the assoc list, plus string versions of the dir defines ("[NORTH]" == "1") already exist in the string table, they're used in a few places, so the lookups "[dir]" should be pretty fast.

Reply:
The defines exist in space_transitions.dm, move them up to __DEFINES/maps.dm